airbyte-integrations/bases/standard-destination-test/src/main/java/io/airbyte/integrations/standardtest/destination/DestinationAcceptanceTest.java

```diff
@@ -1336,22 +1340,30 @@ public void testStressPerformance() throws Exception {
     final Map<String, String> fieldFormats = new HashMap<>();
 
     streams.stream().map(AirbyteStream::getJsonSchema).forEach(streamSchema -> {
-      final JsonNode fieldDefinitions = streamSchema.get("properties");
-      final Iterator<Entry<String, JsonNode>> iterator = fieldDefinitions.fields();
-      while (iterator.hasNext()) {
-        Map.Entry<String, JsonNode> entry = iterator.next();
-        if (entry.getValue().has("format")) {
-          String format = entry.getValue().get("format").asText();
-          if (format.equalsIgnoreCase("date") || format.equalsIgnoreCase("date-time")) {
-            fieldFormats.put(entry.getKey(), format);
-          }
-        }
-      }
+      putFieldsFromSchema(streamSchema, fieldFormats);
     });
 
     return fieldFormats;
   }
 
+  private static void putFieldsFromSchema(JsonNode streamSchema, Map<String, String> fieldFormats) {
+    final JsonNode fieldDefinitions = streamSchema.get("properties");
+    final Iterator<Entry<String, JsonNode>> iterator = fieldDefinitions.fields();
+    while (iterator.hasNext()) {
+      Map.Entry<String, JsonNode> entry = iterator.next();
+      if (entry.getValue().has("type") && entry.getValue().get("type").asText().equals("object")
+          && entry.getValue().has("properties")) {
+        putFieldsFromSchema(entry.getValue(), fieldFormats);
+      }
+      if (entry.getValue().has("format")) {
+        String format = entry.getValue().get("format").asText();
+        if (format.equalsIgnoreCase("date") || format.equalsIgnoreCase("date-time")) {
+          fieldFormats.put(entry.getKey(), format);
```

Comment:
does this handle cases where a nested object has the same key as another field? e.g.
```json
{
  "type": "object",
  "properties": {
    "details": {
      "type": "object",
      "properties": {
        "created_at": {
          "type": "string",
          "format": "date"
        }
      }
    },
    "created_at": {
      "type": "string",
      "format": "date"
    }
  }
}
```
example object:
```json
{
  "details": {
    "created_at": "2022-01-23"
  },
  "created_at": "2022-03-24"
}
```

I _think_ it's going to be difficult to use a flat `Map<String, String>` to represent this sort of schema, but maybe I've misunderstood what this method does

Reply:
Yes, it handles that case. I modified the test case to be sure that it will be handled

Reply:
hm. does this handle the opposite case (where they have conflicting types)?
```json
{
  "details": {
    "created_at": "2022-01-23"
  },
  "created_at": 1648075772
}
```

more broadly, I'd appreciate a summary of how everything fits together - this is a pretty big PR so I want to make sure I understand it well. My current understanding:
* `DestinationAcceptanceTest#whateverTestMethod` calls `getDateTimeFieldsFormat` to populate `dateTimeFieldNames` with the schemas of any date/time-formatted fields
* It then calls `convertDateTimeFields` to serialize date/time-formatted fields to epoch millis/days (relying on `dateTimeFieldNames` to find the fields which need to be modified), and this sometimes converts objects into serialized JSON (?)
* Then it calls `deserializeNestedObjects` to deserialize those nested objects (?)
* And then it can assert that those messages are correct

But I'm not sure (a) where that serialized JSON is generated, or (b) why it needs to be serialized

Reply:
Your understanding is mostly correct. Let me clarify some steps:

> It then calls convertDateTimeFields to serialize date/time-formatted fields to epoch millis/days (relying on dateTimeFieldNames to find the fields which need to be modified), and this sometimes converts objects into serialized JSON (?)

Some destinations write date/time formatted in their specific format or epoch millis/days. E.g. the saved value for "2021-01-06T01:01:01-01:00" in Postgres DB will be "2021-01-06T02:01:01Z".To assert these values it should be cast to one common format. `convertDateTimeFields` converts these values relying on dateTimeFieldNames to find the fields which need to be modified. 
`convertDateTimeFields` doesn't covert objects into serialized JSON. Serialized JSON objects already come from database because the destination stores nested objects as jsonb. And `DestinationAcceptanceTest#retrieveNormalizedRecords` returns messages with serialized nested json objects. Example of how it is saved in Postgres:
![Screenshot from 2022-03-25 02-44-41](https://user-images.githubusercontent.com/35812734/160033026-9b27095b-507e-4035-9ce9-2c92442ccf11.png)
![Screenshot from 2022-03-25 02-44-51](https://user-images.githubusercontent.com/35812734/160033034-f0604536-7cbd-49cc-ba47-90cc0ad29f1c.png)

`deserializeNestedObjects` deserializes these JSONs to JsonNode for future assertion.

Reply:
`hm. does this handle the opposite case (where they have conflicting types)?`
Added additional verification for this case. It will convert only those values which have a date/time format. Also added a test case for that

Reply:
> additional verification for this case

Do both these cases work?
```json
{
  "details": {
    "created_at": "2022-01-23"
  },
  "created_at": "2022-01-23T01:01:01-01:00"
}
```
```json
{
  "details": {
    "created_at": "2022-01-23T01:01:01-01:00"
  },
  "created_at": "2022-01-23"
}
```

If not: I would suggest using a different approach that doesn't only rely on the field name (`"created_at"`) as the map key. I think even if we add special handling for _this_ specific example, there might still be other situations that I can't think of which would need even more special handling.

Reply:
I refactored the approach to not use field path (`"/details/created_at"`) instead of fields name. It gives confidence that the method converts only defined fields with date-time formats. If the destination writes nested objects as jsonb (e.g. Posgtgres, MySQL etc.) it will use a similar approach to the previous one.